File: scripts/ledger/restaurants_and_cafes.py
#!/usr/bin/python3

import subprocess
import csv
import os
from datetime import date, timedelta
from decimal import Decimal, getcontext
import argparse
import json

parser = argparse.ArgumentParser()
parser.add_argument("-n", "--number", default=25)
parser.add_argument("-f", "--file", default="main.ledger")
args = parser.parse_args()
number = int(args.number)

getcontext().prec = 6

year_ago = date.today() - timedelta(days=365)

command=[
    'hledger',
    '-f', args.file,
    'register',
    'expenses:Food:Restaurants & Cafes$',
    '--value', 'then,SEK',
    '--output-format', 'csv',
    '--begin', year_ago.isoformat(),
]

output = subprocess.run(command,check=True, stdout=subprocess.PIPE, universal_newlines=True, encoding='utf-8')
transactions = csv.reader(output.stdout.splitlines()[1:])

def convert(row):
    return {
        'date': date.fromisoformat(row[1]),
        'payee': row[3].split('|')[0].strip(),
        'amount': Decimal(row[5].split(' ')[0]),
        'currency': row[5].split(' ')[1]
    }

transactions = map(convert, transactions)

# calculate top 50 payees by amount sum, and print them like this as json:
# payee, total amount, currency, number of transactions
def groupby(transactions, get_key):
    groups = {}
    for transaction in transactions:
        key = get_key(transaction)
        if key not in groups:
            groups[key] = {
                'payee': transaction['payee'],
                'amount': Decimal(0),
                'currency': transaction['currency'],
                'count': 0
            }
        groups[key]['amount'] += transaction['amount']
        groups[key]['count'] += 1
    return groups

groups = groupby(transactions, lambda t: t['payee'])

result = map(lambda key: groups[key] , sorted(groups, key=lambda g: groups[g]['count'], reverse=True)[:number])

class DecimalEncoder(json.JSONEncoder):
  def default(self, obj):
    if isinstance(obj, Decimal):
      return format(obj, "2,f").replace(',', ' ').replace('.', ',')
      
    return json.JSONEncoder.default(self, obj)

script_location = os.path.realpath(__file__)
script_dir = os.path.dirname(script_location)
dst = os.path.join(script_dir, "../../data/restaurants_and_cafes.json")
dst = os.path.normpath(dst)
content = json.dumps(list(result), indent=4, cls=DecimalEncoder, ensure_ascii=False )
with open(dst, 'w') as f:
    f.write(content)

File: scripts/ledger/restaurants-and-cafes.ts
import { spawn } from 'child_process';
import csv from '@fast-csv/parse';
import { startOfDay } from 'date-fns';
import currency from 'currency.js';
import { addYears } from 'date-fns';
import yargs from 'yargs';
import { writeJSON } from '../utils.js';
import locations from './locations.js';

const argv = yargs(process.argv.slice(2))
	.usage('Usage: $0 <command> [options]')
	.option('file', {
		alias: 'f',
		type: 'string',
		description: 'The ledger file to use'
	})
	.option('number', {
		alias: 'n',
		type: 'number',
		description: 'Minimum number of visits'
	})
	.option('output', {
		alias: 'o',
		type: 'string',
		description: 'Output file',
		demandOption: true
	})
	.parseSync();

type Transaction = { date: Date; amount: currency; payee: string; currency: string };

const log = (...args: any[]) => {
	console.log(...args);
	return args;
};

const hledger = (args: string[]): Promise<any[]> =>
	new Promise((resolve, reject) => {
		const rows = [];
		const parseCSV = csv
			.parse({ headers: true })
			.on('error', reject)
			.on('data', (row) => rows.push(row))
			.on('end', () => resolve(rows));
		const child = spawn('hledger', [...args]);
		let error = '';
		child.stderr.on('data', (data) => (error += data));
		child.stderr.on('end', () => {
			if (error) {
				reject(error);
			}
		});
		child.stdout.pipe(parseCSV);
	});

const exportTransactions = async (file?: string) => {
	log('exporting transactions...');
	const withFile = file ? ['--file', file] : [];
	const params = [
		'register',
		'--value=then,SEK',
		'--output-format=csv',
		'expenses:Food:Restaurants & Cafes$',
		'--begin',
		addYears(new Date(), -1).toISOString().slice(0, 10)
	];
	return await hledger(params.concat(withFile)).then((rows) =>
		rows.map((row: any): Transaction => {
			const { date, description, amount } = row;
			const amountParts = amount.split(' ');
			return {
				date: startOfDay(new Date(date)),
				payee: description.split('|')[0],
				amount: currency(amountParts[0]),
				currency: amountParts[1]
			};
		})
	);
};

export const groupBy = <T, K extends keyof any>(list: T[], getKey: (item: T) => K) =>
	list.reduce((previous, currentItem) => {
		const group = getKey(currentItem);
		if (!previous[group]) previous[group] = [];
		previous[group].push(currentItem);
		return previous;
	}, {} as Record<K, T[]>);

exportTransactions(argv.file)
	.then((transactions) =>
		Object.entries(groupBy(transactions, (t: Transaction) => t.payee))
			.map(([payee, transactions]) => ({
				payee,
				// filter out negative transactions as they are refunds
				count: transactions.filter((t) => t.amount.intValue > 0).length,
				currency: transactions[0].currency,
				amount: transactions.reduce((sum, t) => sum.add(t.amount), currency(0))
			}))
			.filter((e) => e.count >= argv.number)
			.sort((a, b) => b.count - a.count)
			.map((entry) => {
				const payee = entry.payee.trim();
				const location = locations[payee];
				if (!location) throw new Error(`${entry.payee}: location missing`);
				return { ...entry, location };
			})
	)
	.then(writeJSON(argv.output))
	.then(() => console.log('done!'));
